File: src/main/java/io/vproxy/vfx/ui/scene/VSceneRole.java
package io.vproxy.vfx.ui.scene;

import static io.vproxy.vfx.ui.scene.VSceneRoleConsts.*;

class VSceneRoleConsts {
    static final int BIT_MANAGE_WIDTH = 0x0001;
    static final int BIT_MANAGE_HEIGHT = 0x0002;
    static final int BIT_SHOW_COVER = 0x0004;
    static final int BIT_TEMPORARY = 0x008;
    static final int BIT_CENTRAL_WIDTH = 0x010;
    static final int BIT_CENTRAL_HEIGHT = 0x020;

    private VSceneRoleConsts() {
    }
}

public enum VSceneRole {
    MAIN(BIT_MANAGE_WIDTH | BIT_MANAGE_HEIGHT),
    TEMPORARY(BIT_MANAGE_WIDTH | BIT_MANAGE_HEIGHT | BIT_TEMPORARY),
    DRAWER_VERTICAL(BIT_MANAGE_HEIGHT | BIT_SHOW_COVER | BIT_TEMPORARY),
    DRAWER_HORIZONTAL(BIT_MANAGE_WIDTH | BIT_SHOW_COVER | BIT_TEMPORARY),
    POPUP(BIT_SHOW_COVER | BIT_TEMPORARY | BIT_CENTRAL_WIDTH | BIT_CENTRAL_HEIGHT),
    ;
    public final boolean manageWidth;
    public final boolean manageHeight;
    public final boolean showCover;
    public final boolean temporary;
    public final boolean centralWidth;
    public final boolean centralHeight;

    VSceneRole(int bits) {
        this.manageWidth = (bits & BIT_MANAGE_WIDTH) == BIT_MANAGE_WIDTH;
        this.manageHeight = (bits & BIT_MANAGE_HEIGHT) == BIT_MANAGE_HEIGHT;
        this.showCover = (bits & BIT_SHOW_COVER) == BIT_SHOW_COVER;
        this.temporary = (bits & BIT_TEMPORARY) == BIT_TEMPORARY;
        this.centralWidth = (bits & BIT_CENTRAL_WIDTH) == BIT_CENTRAL_WIDTH;
        this.centralHeight = (bits & BIT_CENTRAL_HEIGHT) == BIT_CENTRAL_HEIGHT;
    }
}

File: src-ts/src/ui/scene/VSceneRole.ts
export enum VSceneRole {
  MAIN = 'MAIN',
  TEMPORARY = 'TEMPORARY',
  DRAWER_VERTICAL = 'DRAWER_VERTICAL',
  DRAWER_HORIZONTAL = 'DRAWER_HORIZONTAL',
  POPUP = 'POPUP',
}

interface RoleAttrs {
  manageWidth: boolean;
  manageHeight: boolean;
  showCover: boolean;
  temporary: boolean;
  centralWidth: boolean;
  centralHeight: boolean;
}

const ROLE_ATTRS: Record<VSceneRole, RoleAttrs> = {
  [VSceneRole.MAIN]: {
    manageWidth: true, manageHeight: true,
    showCover: false, temporary: false,
    centralWidth: false, centralHeight: false,
  },
  [VSceneRole.TEMPORARY]: {
    manageWidth: true, manageHeight: true,
    showCover: false, temporary: true,
    centralWidth: false, centralHeight: false,
  },
  [VSceneRole.DRAWER_VERTICAL]: {
    manageWidth: false, manageHeight: true,
    showCover: true, temporary: true,
    centralWidth: false, centralHeight: false,
  },
  [VSceneRole.DRAWER_HORIZONTAL]: {
    manageWidth: true, manageHeight: false,
    showCover: true, temporary: true,
    centralWidth: false, centralHeight: false,
  },
  [VSceneRole.POPUP]: {
    manageWidth: false, manageHeight: false,
    showCover: true, temporary: true,
    centralWidth: true, centralHeight: true,
  },
};

export function roleAttrs(r: VSceneRole): RoleAttrs {
  return ROLE_ATTRS[r];
}
